# Orders Consistency Checks

### The following will be discussed in this script:

1. Checking Dataset
2. Mixed-Type Data
3. Missing Values
4. Duplicate Values
5. Exporting Dataframe

In [1]:
# Importing Libraries 

import pandas as pd
import numpy as np
import os 

In [2]:
path = r'/Users/tsique/Documents/Instacart Basket Analysis'

In [5]:
# Importing orders_wrangled_csv
df_ords = pd.read_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_wrangled.csv'), index_col = 0)

# 01. Checking Dataset

In [6]:
# Checking first 5 and last 5 rows 

df_ords

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3421083 rows x 6 columns]

In [7]:
# Checking basic stats

df_ords.describe()

order_id       user_id  order_number  orders_day_of_week  \
count  3.421083e+06  3.421083e+06  3.421083e+06        3.421083e+06   
mean   1.710542e+06  1.029782e+05  1.715486e+01        2.776219e+00   
std    9.875817e+05  5.953372e+04  1.773316e+01        2.046829e+00   
min    1.000000e+00  1.000000e+00  1.000000e+00        0.000000e+00   
25%    8.552715e+05  5.139400e+04  5.000000e+00        1.000000e+00   
50%    1.710542e+06  1.026890e+05  1.100000e+01        3.000000e+00   
75%    2.565812e+06  1.543850e+05  2.300000e+01        5.000000e+00   
max    3.421083e+06  2.062090e+05  1.000000e+02        6.000000e+00   

       order_hour_of_day  days_since_prior_order  
count       3.421083e+06            3.214874e+06  
mean        1.345202e+01            1.111484e+01  
std         4.226088e+00            9.206737e+00  
min         0.000000e+00            0.000000e+00  
25%         1.000000e+01            4.000000e+00  
50%         1.300000e+01            7.000000e+00  
75%         1.600000e+01            1.500000e+01  
max         2.300000e+01            3.000000e+01

# 02. Mixed -Type Data

In [8]:
# Checking for mixed-type 

for col in df_ords.columns.tolist():
    weird = (df_ords[[col]].applymap(type) != df_ords[[col]].iloc[0].apply(type)).any(axis = 1)
    if len (df_ords[weird]) > 0:
        print (col)

### Nothing was printed, so the statement is true and no mixed-type data has been found in the dataset

# 03. Missing Values

In [9]:
# Checking for missing values in the dataset

df_ords.isnull().sum()

order_id                       0
user_id                        0
order_number                   0
orders_day_of_week             0
order_hour_of_day              0
days_since_prior_order    206209
dtype: int64

### In days_since_prior_order there are 206,209 missing values

In [10]:
# Creating a new dataframe, df_nan, of the missing values 

df_nan = df_ords[df_ords['days_since_prior_order'].isnull() == True]

In [11]:
# Checking df_nan was created properly

df_nan

order_id  user_id  order_number  orders_day_of_week  \
0         2539329        1             1                   2   
11        2168274        2             1                   2   
26        1374495        3             1                   1   
39        3343014        4             1                   6   
45        2717275        5             1                   3   
...           ...      ...           ...                 ...   
3420930    969311   206205             1                   4   
3420934   3189322   206206             1                   3   
3421002   2166133   206207             1                   6   
3421019   2227043   206208             1                   1   
3421069   3154581   206209             1                   3   

         order_hour_of_day  days_since_prior_order  
0                        8                     NaN  
11                      11                     NaN  
26                      14                     NaN  
39                      11                     NaN  
45                      12                     NaN  
...                    ...                     ...  
3420930                 12                     NaN  
3420934                 18                     NaN  
3421002                 19                     NaN  
3421019                 15                     NaN  
3421069                 11                     NaN  

[206209 rows x 6 columns]

In [12]:
# Finding out the total number of rows in the data 

df_ords.shape

(3421083, 6)

## 6% of the data is missing 

## Addressing Missing Values

In [13]:
# Creating a dataframe without the missing values

df_ords_clean = df_ords[df_ords['days_since_prior_order'].isnull() == False]

In [14]:
df_ords_clean

order_id  user_id  order_number  orders_day_of_week  \
1         2398795        1             2                   3   
2          473747        1             3                   3   
3         2254736        1             4                   4   
4          431534        1             5                   4   
5         3367565        1             6                   2   
...           ...      ...           ...                 ...   
3421078   2266710   206209            10                   5   
3421079   1854736   206209            11                   4   
3421080    626363   206209            12                   1   
3421081   2977660   206209            13                   1   
3421082    272231   206209            14                   6   

         order_hour_of_day  days_since_prior_order  
1                        7                    15.0  
2                       12                    21.0  
3                        7                    29.0  
4                       15                    28.0  
5                        7                    19.0  
...                    ...                     ...  
3421078                 18                    29.0  
3421079                 10                    30.0  
3421080                 12                    18.0  
3421081                 12                     7.0  
3421082                 14                    30.0  

[3214874 rows x 6 columns]

## Creating a new dataframe we do not risk losing data through deleting and using a mean or median would not be accurate for later calculations.

# 04. Duplicate Values

In [15]:
# Creating a subset of only rows that are duplicates in df_ords_clean

df_dups = df_ords_clean[df_ords_clean.duplicated()]

In [16]:
df_dups

Empty DataFrame
Columns: [order_id, user_id, order_number, orders_day_of_week, order_hour_of_day, days_since_prior_order]
Index: []

## It looks as though there are no duplicated in the clean data

# 05. Exporting dataframe

In [17]:
df_ords_clean.to_csv(os.path.join(path,'02 Data', 'Prepared Data', 'orders_checked.csv'))